atoma-state/src/state_manager.rs

```diff
@@ -325,47 +341,229 @@ impl AtomaState {
+    /// - The maximum number of compute units supported
     ///
-    /// This function will return an error if the database query fails.
+    /// # Security
+    ///
+    /// When `is_confidential` is true, this method:
+    /// - Only returns nodes that have valid public keys
+    /// - Only considers tasks with security level 2
+    /// - Requires nodes to be registered in the node_public_keys table
+    ///
+    /// # Example
+    ///
+    /// ```rust,ignore
+    /// # use atoma_state::AtomaState;
+    /// # use sqlx::PgPool;
+    ///
+    /// async fn find_cheapest_node(state: &AtomaState) -> anyhow::Result<()> {
+    ///     // Find cheapest non-confidential node for GPT-4
+    ///     let regular_node = state.get_cheapest_node_for_model("gpt-4", false).await?;
+    ///     
+    ///     // Find cheapest confidential node for GPT-4
+    ///     let confidential_node = state.get_cheapest_node_for_model("gpt-4", true).await?;
+    ///     
+    ///     Ok(())
+    /// }
+    /// ```
     #[instrument(level = "trace", skip_all, fields(%model))]
-    pub async fn get_cheapest_node_for_model(&self, model: &str) -> Result<Option<CheapestNode>> {
-        let node_settings = sqlx::query(
-            "SELECT tasks.task_small_id, price_per_compute_unit, max_num_compute_units 
-            FROM (SELECT * 
-                  FROM tasks 
-                  WHERE is_deprecated=false
-                    AND model_name = $1) AS tasks 
-            JOIN (SELECT * 
-                  FROM node_subscriptions
-                  WHERE valid = true) AS node_subscriptions 
-            ON tasks.task_small_id=node_subscriptions.task_small_id 
+    pub async fn get_cheapest_node_for_model(
+        &self,
+        model: &str,
+        is_confidential: bool,
+    ) -> Result<Option<CheapestNode>> {
+        let mut query = String::from(
+            r#"
+            SELECT tasks.task_small_id, price_per_compute_unit, max_num_compute_units, node_subscriptions.node_small_id
+            FROM tasks
+            INNER JOIN node_subscriptions ON tasks.task_small_id = node_subscriptions.task_small_id"#,
+        );
+
+        if is_confidential {
+            query.push_str(r#"
+            INNER JOIN node_public_keys ON node_public_keys.node_small_id = node_subscriptions.node_small_id"#);
+        }
+
+        query.push_str(
+            r#"
+            WHERE tasks.is_deprecated = false
+            AND tasks.model_name = $1
+            AND node_subscriptions.valid = true"#,
+        );
+
+        if is_confidential {
+            query.push_str(
+                r#"
+            AND tasks.security_level = 2
+            AND node_public_keys.is_valid = true"#,
+            );
+        }
+
+        query.push_str(
+            r#"
             ORDER BY node_subscriptions.price_per_compute_unit 
-            LIMIT 1",
+            LIMIT 1"#,
+        );
+
+        let node_settings = sqlx::query(&query)
+            .bind(model)
+            .bind(false)
```

Comment:
the 2nd bind doesn have $2 in the query string

Reply:
Oh good catch !